### PR TITLE
bluestore, NVMEDevice: Add multiple thread support while doing SPDK NVMe I/O

### PR DESCRIPTION
**Pull request by optimistyzy**
Previously, we only have one thread to do SPDK I/O, and this patch
adds the multiple thread support.  

Signed-off-by: Ziye Yang <optimistyzy@gmail.com>

### CODE REVIEW

File: src/os/bluestore/NVMEDevice.cc [hunk 1]
```diff
@@ -51,14 +51,14 @@
...
 
-std::vector<void*> data_buf_mempool;
-
 static constexpr uint16_t data_buffer_default_num = 2048;
 
 static constexpr uint32_t data_buffer_size = 8192;
 
 static constexpr uint16_t inline_segment_num = 32;
 
+static __thread int queue_id = -1;
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: thread_local in c++11

File: src/os/bluestore/NVMEDevice.cc [hunk 2]
```diff
@@ -73,7 +73,7 @@ enum {
   l_bluestore_nvmedevice_last
 };
 
-static void io_complete(void *t, const struct spdk_nvme_cpl *completion);
+void io_complete(void *t, const struct spdk_nvme_cpl *completion);
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: why remove static?

File: src/os/bluestore/NVMEDevice.cc [hunk 3]
```diff
@@ -231,24 +157,123 @@ class SharedDriverData {
...
+  }
+
+  ~SharedDriverQueueData() {
     g_ceph_context->get_perfcounters_collection()->remove(logger);
     if (!qpair) {
       spdk_nvme_ctrlr_free_io_qpair(qpair); 
     }
     delete logger;
   }
+};
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: blank line needed

File: src/os/bluestore/NVMEDevice.cc (hunk 1 above)
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: instead of owning queue_id, we can directly own Queue*

Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: It seems that for object can we use thread_local define that ?

**Comment by optimistyzy** [comment]
Can we thread_local before object variable pointer?

**Comment by optimistyzy** [comment]
This patch needs to be verified for perf

**Pull request closed without merging**